torch/csrc/autograd/function.h

```diff
@@ -257,11 +283,62 @@ struct TORCH_API Node : std::enable_shared_from_this<Node> {
   // Miscellaneous Methods
   //~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 
-  /// The sequence number of this `Node`.
+  /// NOTE [ Sequence Number]
+  ///
+  /// The sequence_nr has two main usages in autograd:
+  ///
+  /// 1) Determine the node's execution priority in the engine.
+  ///    Nodes with higher priority numbers are executed first. There is actually no
+  ///    particular reason why we want nodes corresponding to ops executed later to be
```

Comment:
Actually there is.
We want to make sure that code like this runs without using too much memory:
```
inp = torch.rand(10)
out1 = utils.checkpoint(inp)
out2 = utils.checkpoint(inp)
(out1 + out2).sum().backward()
```

The important part here is that once you enter a branch with a checkpoint, you do re-entrant autograd with a graph you just created. And you actually want to execute all that re-entrant graph before executing the graph from the rest of the model.
Here it would be bad because it would defeat all the memory saving from the checkpoint.

Reply:
Interesting. This is only because of reentrant depth though right? I can see why order is important now (for the checkpointing case), but seems like sequence numbers don't necessarily play a role in that.

Reply:
The other more general idea is really that the backward is supposed to run the forward in reversed order. So it is natural to run ops in the reversed order if possible.